### PR TITLE
Replace OSError exceptions from `can_read` with `redis.ConnectionError`

### LINKED ISSUE #2138: `connection.can_read()` raises OSError when connection is broken
**Version**: 4.2.2

**Platform**: Windows 11
**Description**: 
`redis.asyncio.connection.Connection()`, the class provides a `can_read` method, which is used, for example, in `Client.parse_response`.

However, unlike `Connection.read_response()` there is no error handling present to turn any OSErrors into `asyncio.ConnectionError`.
And so, the OSError bubbles up and causes problems, where we expect only `asyncio.ConnectionError` to occur.
This plays havoc with retry-mechanisms and automatic reconnect.

The issue is evident when the connection to the redis server is interrupted during a blocking call to `pubsub.run()` where a `get_message` call is made with a positive timeout.

### PR DESCRIPTION
**Pull request by kristjanvalur**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

This change catches OS errors in `Connection.can_read`, closes connection and raises a `redis.ConnectionError`.
Fixes #2138

### CODE REVIEW

**Comment by chayim** [comment]
@kristjanvalur I tried to push an update to your branch, to fix the linters. Can you run the following to fix your linting? Then we can merge!

```
black redis
```

File: redis/asyncio/connection.py [hunk 1]
```diff
@@ -911,7 +911,13 @@ async def can_read(self, timeout: float = 0):
         """Poll the socket to see if there's data that can be read."""
         if not self.is_connected:
             await self.connect()
-        return await self._parser.can_read(timeout)
+        try:
+            return await self._parser.can_read(timeout)
+        except OSError as e:
+            await self.disconnect()
+            raise ConnectionError(
+                f"Error while reading from {self.host}:{self.port} : {e.args}"
```
Comment (chayim) [inline redis/asyncio/connection.py:919]: Nice. Small nit to remove a space.
```
                f"Error while reading from {self.host}:{self.port}: {e.args}"
```

File: redis/connection.py [hunk 2]
```diff
@@ -808,7 +808,11 @@ def can_read(self, timeout=0):
         sock = self._sock
         if not sock:
             self.connect()
-        return self._parser.can_read(timeout)
+        try:
+            return self._parser.can_read(timeout)
+        except OSError as e:
+            self.disconnect()
+            raise ConnectionError(f"Error while reading from {self.host}:{self.port} : {e.args}")
```
Comment (chayim) [inline redis/connection.py:815]: Same
```
                f"Error while reading from {self.host}:{self.port}: {e.args}"
```

**Review by chayim — CHANGES_REQUESTED** [review]
Ready to merge pending the lint fix, and this small change!

**Review by dvora-h — APPROVED** [review]

**Review by chayim — APPROVED** [review]

**Pull request merged**